File: cpp/day01/src/main.cxx
#include <fstream>
#include <iostream>
#include <ranges>
#include "lib.hpp"

using namespace std;

int main(int argc, char **argv) {
  if (argc != 2) {
    cout << "Invalid input parameters. Usage: part1 <file path>" << endl;
  }
  char *filename = argv[1];

  ifstream is(filename);
  vector<int> left, right;
  for (string line; getline(is, line);) {
    auto items = line_to_numbers(line);
    left.push_back(get<0>(items));
    right.push_back(get<1>(items));
  }
  sort(left.begin(), left.end());
  sort(right.begin(), right.end());
  int result = 0;
  for (size_t idx = 0; idx < left.size(); idx++) {
    result += abs(left[idx] - right[idx]);
  }
  cout << result << endl;
  return 0;
}

File: rust/day01/src/main.rs
use std::{env, fs};

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Require input file path");
    }

    let file_path = &args[1];
    let content = fs::read_to_string(file_path).expect("cannot read file");
    let mut left: Vec<i32> = vec![];
    let mut right: Vec<i32> = vec![];
    for line in content.trim().split('\n') {
        let items: Vec<&str> = line.split(' ').collect();
        left.push(items.first().expect("invalid input").parse().unwrap());
        right.push(items.last().expect("invalid input").parse().unwrap());
    }
    left.sort();
    right.sort();
    let mut result: i32 = 0;
    for (idx, val) in left.iter().enumerate() {
        result += (val - right[idx]).abs();
    }
    println!("{result}");
}
